include/pstl_offload/internal/usm_memory_replacement_common.h

```diff
@@ -0,0 +1,130 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_PSTL_OFFLOAD_INTERNAL_USM_MEMORY_REPLACEMENT_COMMON_H
+#define _ONEDPL_PSTL_OFFLOAD_INTERNAL_USM_MEMORY_REPLACEMENT_COMMON_H
+
+#include <sycl/sycl.hpp>
+#include <cstddef>
+#include <cstdlib>
+#include <cassert>
+#include <cstring>
+
+#if __linux__
+#include <dlfcn.h>
+#include <unistd.h>
+#endif // __linux__
+
+namespace __pstl_offload {
+
+inline constexpr std::size_t __uniq_type_const = 0x23499abc405a9bccLLU;
+
+struct __block_header {
+    std::size_t _M_uniq_const;
+    void* _M_original_pointer;
+    std::size_t _M_requested_number_of_bytes;
+    sycl::device* _M_device;
+}; // struct __block_header
+
+inline constexpr std::size_t __header_offset = 32;
+
+static_assert(sizeof(__block_header) <= __header_offset);
+
+#if __linux__
+
+inline std::size_t __get_memory_page_size()
+{
+    static std::size_t __memory_page_size = sysconf(_SC_PAGESIZE);
+    return __memory_page_size;
+}
+
+inline void* __allocate_from_device(sycl::device* __device, std::size_t __size, std::size_t __alignment) {
```

Comment:
Memory from device is allocated by `sycl::*lloc_device()` family, at this place we use `sycl::aligned_alloc_shared()`, so might `__allocate_from_device()` be renamed somehow to closely match kind of allocated memory?

Reply:
Renamed to `__allocate_shared_for_device`.